src/plugins/data/server/search/aggs/aggs_service.ts

```diff
@@ -0,0 +1,127 @@
+  /**
+   * This is a temporary cache for agg types which are registered by other
+   * plugins in `setup`. Its contents are merged with the default agg types
+   * and added to the "real" registry which is instantiated when someone calls
+   * `data.search.aggs.asScopedToClient`.
+   */
+  private readonly aggTypesCache = new AggTypesRegistry();
+
+  /**
+   * getForceNow uses window.location on the client, so we must have a
+   * separate implementation of calculateBounds on the server.
+   */
+  private calculateBounds = (timeRange: TimeRange) => calculateBounds(timeRange, {});
+
+  public setup({ registerFunction }: AggsSetupDependencies): AggsSetup {
+    this.registerFunction = registerFunction;
+
+    return {
+      types: this.aggTypesCache.setup(),
+    };
+  }
+
+  public start({ fieldFormats, uiSettings }: AggsStartDependencies): AggsStart {
+    return {
+      asScopedToClient: async (savedObjectsClient: SavedObjectsClientContract) => {
+        const uiSettingsClient = uiSettings.asScopedToClient(savedObjectsClient);
+        const formats = await fieldFormats.fieldFormatServiceFactory(uiSettingsClient);
+
+        // cache ui settings, only including items which are explicitly needed by aggs
+        const uiSettingsCache = pick(await uiSettingsClient.getAll(), aggsRequiredUiSettings);
+        const getConfig = <T = any>(key: string): T => {
+          return uiSettingsCache[key];
+        };
+
+        const defaultAggTypes = getAggTypes({
+          calculateBounds: this.calculateBounds,
+          getConfig,
+          getFieldFormatsStart: () => ({
+            deserialize: formats.deserialize,
+            getDefaultInstance: formats.getDefaultInstance,
+          }),
+          /**
+           * Date histogram and date range need to know whether we are using the
+           * default timezone, but `isDefault` is not currently offered on the
+           * server, so we need to manually check for the default value.
+           */
+          isDefaultTimezone: () => getConfig('dateFormat:tz') === 'Browser',
+        });
+
+        /**
+         * Instantiating a new `AggsCommonService` will create a new registry, which
+         * is desirable in this case as the scoped ui settings need to be provided
+         * to some of the registered default aggs. So we merge the cached agg types
+         * from `setup` with the default aggs that use scoped settings, ensuring
+         * downstream consumers are receiving everything that has been registered.
+         */
+        const aggsCommonService = new AggsCommonService();
+        const cachedAggTypes = this.aggTypesCache.start().getAll();
+
+        // call setup to ensure all default agg types & functions are registered
+        aggsCommonService.setup({
+          aggTypes: {
+            buckets: defaultAggTypes.buckets.concat(cachedAggTypes.buckets),
+            metrics: defaultAggTypes.metrics.concat(cachedAggTypes.metrics),
```

Comment:
We need the default agg types which are in the registry to have ui settings which are scoped to the current user. However, technically 3rd party plugins can register their own agg types.

So for now we are making the assumption that 3rd party agg types are _not_ using ui settings and therefore don't require a scoped saved objects client. This means we can store one shared copy of them in `setup`, and then merge everything together here.

I think this is fine for now since we don't even have known 3rd party plugins registering custom agg types right now, and they'd need to be doing it from the server anyway.

Longer term we should consider revisiting the design of the agg types and perhaps simplify things so that we could do something like give each agg type its own `getConfig` as a convenience, which could be used inside of the agg. But this would likely require a simpler "definition" style of registry item, similar to what we've done elsewhere (like expressions).